### PR TITLE
refactor sectionProgressRedux

### PR DESCRIPTION
**Pull request by cforkish**
this is a revival of the PR that started this whole refactor project #37246 and was extracted from the completed section progress table refactor #38309.

The purpose of this refactor is the separate student-specific progress data from general level/lesson/lessonGroup data. The motivation was that with very large sections, merging all the progress data with level data was resulting in _a lot_ of data bloat (see below). And since progress data comes from a separate api endpoint than script data, it seemed like a logical division.

## sectionProgressRedux
`/dashboardapi/section_level_progress/` returns an object with two top-level properties: 
```
serverData = {
  student_progress: {
    studentId: {
      levelId: {
        status: string,
        result: number,
        paired: bool
      }
    }
  }
  student_last_updates: {
    studentId: number (timestamp)
  }
}
```

`studentTimestampsByScript` has been renamed to `studentLastUpdateByScript`, but otherwise how we are storing `serverData.student_last_updates` remains unchanged.

for the rest of the progress data, here is how the data structure has changed:

BEFORE
with 250 students, `sectionProgressRedux` store was 109,343 lines and 3.9 MB
```
// redux properties
scriptDataByScript: {},
studentLevelProgressByScript: {},
studentLevelPairingByScript: {},
studentTimestampsByScript: {},
studentLevelTimeSpentByScript: {},
levelsByLessonByScript: {}

// status string was merged with general level data from scriptDataByScript[scriptId] into a full level object 
// that included the student-specific status string. this merge was the cause of massive data bloat.
serverData.student_progress[studentId][levelId].status -> levelsByLessonByScript[scriptId][studentId][stageIndex][levelIndex].status

// pairing data, time spent, and results were stored in their own structures
serverData.student_progress[studentId][levelId].paired -> studentLevelPairingByScript[scriptId][studentId][levelId]
serverData.student_progress[studentId][levelId].time_spent -> studentLevelTimeSpentByScript[scriptId][studentId][levelId]
serverData.student_progress[studentId][levelId].result -> studentLevelProgressByScript[scriptId][studentId][levelId]
```

AFTER
with 250 students, `sectionProgressRedux` store is 12,889 lines and 412 KB
```
// redux properties
scriptDataByScript: {},
studentLevelProgressByScript: {},
studentLastUpdateByScript: {}

// progress data is minimally post-processed to convert snake_case to camelCase and fill in missing properties,
// but otherwise the structure from the server is preserved
serverData.student_progress -> studentLevelProgressByScript[scriptId]
{[status], [result], [paired], [time_spent]} -> {status, result, paired, timeSpent}
```

### CODE REVIEW

File: apps/src/redux/sectionDataRedux.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ export const SET_SECTION = 'sectionData/SET_SECTION';
 export const setSection = section => {
   // Sort section.students by name.
   const sortedStudents = section.students.sort((a, b) =>
-    a.name.localeCompare(b.name)
+    a.name.localeCompare(b.name, undefined, {numeric: true})
```
Comment (maureensturgeon) [inline apps/src/redux/sectionDataRedux.js:38]: TIL `localeCompare`, thanks!

File: apps/src/templates/progress/progressHelpers.js [hunk 2]
```diff
@@ -194,3 +206,87 @@ export const processedLevel = level => {
...
+ * @returns {studentLevelProgressType} Our canonical progress shape
+ */
+export const levelProgressFromServer = serverObject => {
+  return {
+    status: serverObject.status || LevelStatus.not_tried,
+    result: getLevelResult(serverObject),
+    paired: serverObject.paired || false,
+    timeSpent: serverObject.time_spent || 0,
+    pages:
+      serverObject.pages_completed && serverObject.pages_completed.length > 1
```
Comment (maureensturgeon) [inline apps/src/templates/progress/progressHelpers.js]: Should this be `serverObject.pages_completed.length > 0`, should there be mapping if there is 1 `pages_complete`?

File: apps/src/templates/progress/progressHelpers.js [hunk 3]
```diff
@@ -194,3 +206,87 @@ export const processedLevel = level => {
...
+              (pageResult && levelProgressFromResult(pageResult)) ||
+              levelProgressWithStatus(LevelStatus.not_tried)
+          )
+        : null
+  };
+};
+
+/**
+ * Given an object from the server with student progress data keyed by level ID,
+ * parse the progress data into our canonical studenLevelProgressType
```
Comment (maureensturgeon) [inline apps/src/templates/progress/progressHelpers.js]: Typo in studenLevelProgressType missing t

File: apps/src/templates/progress/progressHelpers.js [hunk 4]
```diff
@@ -194,3 +206,87 @@ export const processedLevel = level => {
...
+export const processServerStudentProgress = serverStudentProgress => {
+  return _.mapValues(serverStudentProgress, progress =>
+    levelProgressFromServer(progress)
+  );
+};
+
+/**
+ * Given an object from the server with section progress data keyed by student
+ * ID and level ID, parse the progress data into our canonical
+ * studenLevelProgressType
```
Comment (maureensturgeon) [inline apps/src/templates/progress/progressHelpers.js]: Same typo

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx [hunk 5]
```diff
@@ -109,11 +125,11 @@ class VirtualizedDetailView extends Component {
     this.detailView = ref;
   }
 
-  onClickLevel = lessonOfInterest => {
+  onClickLevel(lessonOfInterest) {
```
Comment (maureensturgeon) [inline apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:128]: Out of curiosity, why did you move away from arrow functions and move to binding instead?

File: apps/src/templates/sectionProgress/sectionProgressLoader.js [hunk 6]
```diff
@@ -142,63 +109,29 @@ function processStudentTimestamps(timestamps) {
...
-    throw new Error('Input is invalid');
-  }
-  return studentPairing;
-}
-
 function postProcessDataByScript(scriptData) {
-  if (!scriptData.stages) {
-    return scriptData;
+  // Filter to match scriptDataPropType
+  const filteredScriptData = {
```
Comment (maureensturgeon) [inline apps/src/templates/sectionProgress/sectionProgressLoader.js:114]: Could leverage `_.pick` https://lodash.com/docs/4.17.15#pick

File: apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js [hunk 7]
```diff
@@ -305,19 +305,21 @@ export function getPluggedLessonCompletionStatus(state, lesson) {
...
-    const levelIds = _.map(lesson.levels, 'activeId');
+    const studentIds = Object.keys(levelProgressByScript);
+    const levelIds = _.map(lesson.levels, 'id');
     let numStudentsCompletedLesson = 0;
     let numStudentsInProgressLesson = 0;
     studentIds.forEach(studentId => {
       let numLevelsInLessonCompletedByStudent = 0;
       levelIds.forEach(levelId => {
+        const levelProgress = levelProgressByScript[studentId][levelId];
+        // console.log('progress', studentId, levelId, levelProgress);
```
Comment (maureensturgeon) [inline apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js]: 👀  rogue log

File: apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js [hunk 8]
```diff
@@ -351,6 +353,7 @@ export function fetchStandardsCoveredForScript(scriptId) {
       dataType: 'json',
       url: `/dashboardapi/script_standards/${scriptId}`
     }).then(data => {
+      console.log('standards data', scriptId, data);
```
Comment (maureensturgeon) [inline apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js]: log - and another below

File: apps/test/unit/templates/sectionProgress/VirtualizedDetailViewTest.js [hunk 9]
```diff
@@ -27,15 +27,27 @@ describe('VirtualizedSummaryView', () => {
     stubRedux();
     registerReducers({sectionProgress, scriptSelection, currentUser});
     defaultProps = {
-      levelsByLesson: {
+      levelProgressByStudent: {
```
Comment (maureensturgeon) [inline apps/test/unit/templates/sectionProgress/VirtualizedDetailViewTest.js:30]: Should be able to use `fakeStudentLevelProgress` from progressTestHelpers, once it's merged in

File: apps/test/unit/templates/sectionProgress/VirtualizedSummaryViewTest.js [hunk 10]
```diff
@@ -27,32 +29,32 @@ describe('VirtualizedSummaryView', () => {
     stubRedux();
     registerReducers({sectionProgress, scriptSelection, currentUser});
     defaultProps = {
-      levelsByLesson: {
+      levelProgressByStudent: {
```
Comment (maureensturgeon) [inline apps/test/unit/templates/sectionProgress/VirtualizedSummaryViewTest.js:32]: Another opportunity to use the helper!

**Review by maureensturgeon — APPROVED** [review]
LGTM, it's be great to get other eyes on this as well. It would also be nice to leverage the added `progressTestHelpers` from https://github.com/code-dot-org/code-dot-org/pull/38642 (I'll get that merged in soon)

File: apps/src/templates/progress/progressHelpers.js [hunk 11]
```diff
@@ -120,48 +126,54 @@ export function stageIsAllAssessment(levels) {
...
+ * level id containing objects representing the student's progress in that level
+ * @param {levelType[]} levels An array of the levels in a stage
  * @returns {object} An object with a total count of levels in each of the
  * following buckets: total, completed, imperfect, incomplete, attempted.
  */
-export function summarizeProgressInStage(levelsWithStatus) {
+export function summarizeProgressInStage(studentProgress, levels) {
   // Filter any bonus levels as they do not count toward progress.
-  levelsWithStatus = levelsWithStatus.filter(level => !level.bonus);
+  const filteredLevels = levels.filter(level => !level.bonus);
```
Comment (jamescodeorg) [inline apps/src/templates/progress/progressHelpers.js:188]: (very minor) Would `nonBonusLevels` be clearer than `filteredLevels`?  (I don't feel strongly about it, up to you.)

File: apps/src/templates/progress/progressHelpers.js [hunk 12]
```diff
@@ -194,3 +206,87 @@ export const processedLevel = level => {
...
+    return TestResults.LOCKED_RESULT;
+  }
+  if (serverProgress.readonly_answers) {
+    return TestResults.READONLY_SUBMISSION_RESULT;
+  }
+  if (serverProgress.submitted) {
+    return TestResults.SUBMITTED_RESULT;
+  }
+
+  return serverProgress.result || resultFromStatus(serverProgress.status);
```
Comment (jamescodeorg) [inline apps/src/templates/progress/progressHelpers.js:272]: Could you explain this logic a bit?  Maybe more generally, what does "result" represent and why is this the algorithm that we use to determine it?

File: apps/src/templates/progress/progressTypes.js [hunk 13]
```diff
@@ -54,6 +54,37 @@ export const levelType = PropTypes.shape({
...
+ * See src/util/sharedConstants.LevelStatus.
+ * @property {number} result
+ * A numerical enum of the TestResult a student received for a level.
+ * See src/constants.TestResult.
+ * See src/code-studio/activityUtils.activityCssClass for a mapping to status.
+ * @property {bool} paired
+ * A boolean indicating if a student was paired on a level.
+ * @property {number} timeSpent
+ * An optional value indicating the time a student spent on a level.
+ * @property {array} pages
```
Comment (jamescodeorg) [inline apps/src/templates/progress/progressTypes.js:73]: This property is not in the shape definition below?

File: apps/src/templates/progress/progressHelpers.js [hunk 14]
```diff
@@ -194,3 +206,87 @@ export const processedLevel = level => {
...
+ * Parse a level progress object that we get from the server using either
+ * /api/user_progress or /dashboardapi/section_level_progress into our
+ * canonical studentLevelProgressType shape.
+ * @param {object} serverObject A progress object from the server
+ * @returns {studentLevelProgressType} Our canonical progress shape
+ */
+export const levelProgressFromServer = serverObject => {
+  return {
+    status: serverObject.status || LevelStatus.not_tried,
+    result: getLevelResult(serverObject),
```
Comment (jamescodeorg) [inline apps/src/templates/progress/progressHelpers.js]: Could you elaborate a bit on what status and result mean and how they are used?

File: apps/src/templates/progress/progressHelpers.js (hunk 12 above)
Comment (cforkish) [reply apps/src/templates/progress/progressHelpers.js:272]: see `progressTypes.studentLevelProgressType` for description of these properties and pointers to their "enum" definitions. 

as you know, we have a lot of technical debt here, and throughout the refactor work it's been necessary to strike a balance between auditing the old code for potential tech debt cleanup vs. preserving legacy logic as-is. cleaning up the relationship between `result` and `status` is going to require a deep dive, so for anything pertaining to that, i opted for preserving logic as-is where possible. 

previously, when processing section progress data in `sectionProgressLoader`, we would pass all the result values returned from the server through [progressRedux.getLevelResult](https://github.com/code-dot-org/code-dot-org/blob/3adbeb6b5ec6227c9d5f114d4ee93cbcd6cc683e/apps/src/code-studio/progressRedux.js#L286), presumably so that the first three conditional blocks could potentially override a server result value. i never dug into why (or even if) that part of this "algorithm" is necessary.

`getLevelResult` was moved into this module as part of the `progressRedux` refactor that got punted, and was expanded to add `|| resultFromStatus` to handle part of said refactor. leaving this function in two places was an oversight when i extracted this work into its own PR. i will update `progressRedux` to use this version.

File: apps/src/templates/progress/progressHelpers.js (hunk 14 above)
Comment (cforkish) [reply apps/src/templates/progress/progressHelpers.js]: see the type definition in `progressTypes.studentLevelProgressType` and let me know if the comments there require elaboration.

File: apps/src/templates/progress/progressHelpers.js (hunk 2 above)
Comment (cforkish) [reply apps/src/templates/progress/progressHelpers.js]: nope. "pages" is only used by multi-page assessments, and its presence (or absence) is use to indicate that level type. i will add a comment explaining that!

File: apps/src/templates/progress/progressHelpers.js (hunk 3 above)
Comment (cforkish) [reply apps/src/templates/progress/progressHelpers.js]: thanks!

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx (hunk 5 above)
Comment (cforkish) [reply apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:128]: so glad you asked! https://medium.com/@charpeni/arrow-functions-in-class-properties-might-not-be-as-great-as-we-think-3b3551c440b1

Comment (cforkish) [reply apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:128]: FYI, see [here](https://github.com/code-dot-org/code-dot-org/pull/38772#discussion_r565761973). you're welcome to get in on the best-practices fun with Jessie, Maddie, and I once i circle back to it if you're interested!

File: apps/src/templates/sectionProgress/sectionProgressLoader.js (hunk 6 above)
Comment (cforkish) [reply apps/src/templates/sectionProgress/sectionProgressLoader.js:114]: i recently ran into a case where i was refactoring some code that leveraged `pick` and found it cumbersome, since it only handles "picking" but not "mapping", so as soon as i started needing anything other than a direct mapping for a property i had to break it out from the `pick` call and things started getting messier. in this case, we'd already need to break out `stages: scriptData.lessons`, and imo this just feels cleaner (and more maintainable) than starting with a `pick` call then handling custom mappings afterward.

if we lived in a world where server shapes === frontend shapes i think `pick` would be much more useful, but in our case i don't think it's a helpful pattern to introduce/encourage.

File: apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js (hunk 7 above)
Comment (cforkish) [reply apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js]: thanks! as you can tell, i had some trouble fixing tests for this module ;-)

File: apps/test/unit/templates/sectionProgress/VirtualizedDetailViewTest.js (hunk 9 above)
Comment (cforkish) [reply apps/test/unit/templates/sectionProgress/VirtualizedDetailViewTest.js:30]: these tests are going away once we finish merging our refactor! :)

File: apps/src/templates/progress/progressTypes.js (hunk 13 above)
Comment (cforkish) [reply apps/src/templates/progress/progressTypes.js:73]: it's in the complete "type" definition, but the property is recursive so can't be included in the initial shape definition. i added it in as a comment, and did the same for `LevelType.subleves` above. let me know if this is clearer.

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx (hunk 5 above)
Comment (maureensturgeon) [reply apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:128]: Yeah, I'd love to!

File: apps/src/templates/progress/progressHelpers.js [hunk 15]
```diff
@@ -120,48 +126,54 @@ export function stageIsAllAssessment(levels) {
...
-        statusCounts.incomplete = statusCounts.incomplete + 1;
-        statusCounts.attempted = statusCounts.attempted + 1;
+        statusCounts.incomplete++;
+        statusCounts.attempted++;
         break;
       case LevelStatus.passed:
-        statusCounts.imperfect = statusCounts.imperfect + 1;
+        statusCounts.imperfect++;
         break;
       // All others are assumed to be not tried
```
Comment (jamescodeorg) [inline apps/src/templates/progress/progressHelpers.js:223]: (very minor) You can move the `LevelStatus.not_tried` case down here if you wanted to make it more explicit that other statuses should be handled as not_tried.

File: apps/src/templates/sectionProgress/sectionProgressConstants.js [hunk 16]
```diff
@@ -18,7 +18,8 @@ export const scriptDataPropType = PropTypes.shape({
     })
   ),
   family_name: PropTypes.string,
-  version_year: PropTypes.string
+  version_year: PropTypes.string,
+  name: PropTypes.string
```
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/sectionProgressConstants.js:22]: (very minor) I don't know if we have a canonical ordering for the props, but `name` feels like it should go higher up in this list, maybe right after id or something.

File: apps/src/templates/sectionProgress/sectionProgressLoader.js [hunk 17]
```diff
@@ -142,63 +109,29 @@ function processStudentTimestamps(timestamps) {
...
-      levelId => typeof studentPairing[userId][levelId] === 'boolean'
-    )
-  );
-  if (!isValid) {
-    throw new Error('Input is invalid');
-  }
-  return studentPairing;
-}
-
 function postProcessDataByScript(scriptData) {
```
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/sectionProgressLoader.js:112]: Maybe rename this to `processScriptData` to be more symmetric with `processServerSectionProgress`?

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx [hunk 18]
```diff
@@ -71,26 +74,34 @@ class VirtualizedDetailView extends Component {
...
-    };
-    this.detailView = null;
     this.setDetailViewRef = this.setDetailViewRef.bind(this);
+    this.onClickLevel = this.onClickLevel.bind(this);
+    this.cellRenderer = this.cellRenderer.bind(this);
+    this.studentCellRenderer = this.studentCellRenderer.bind(this);
+    this.getColumnWidth = this.getColumnWidth.bind(this);
   }
 
+  state = {
```
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:97]: Just curious, why did you move this out of the constructor?

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx [hunk 19]
```diff
@@ -220,10 +236,50 @@ class VirtualizedDetailView extends Component {
...
-  };
+  }
 
-  studentCellRenderer = (studentStartIndex, stageIdIndex, key, style) => {
-    const {section, levelsByLesson, stageExtrasEnabled} = this.props;
+  studentCellRenderer(studentStartIndex, stageIdIndex, key, style) {
+    const {section, levelProgressByStudent, scriptData} = this.props;
+    const student = section.students[studentStartIndex];
+    let child;
+    if (stageIdIndex < 0) {
```
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:245]: What does stageIdIndex < 0 represent?

**Review by jamescodeorg — APPROVED** [review]
Woohoo, congrats on getting this over the finish line!  A few cosmetic comments but looks good to me!

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx (hunk 19 above)
Comment (cforkish) [reply apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:245]: column -1 is the student list (in this legacy component it wasn't a separate component like it is in ProgressTableStudentList in the refactored table). fwiw, this component will be removed once we make the switchover

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx (hunk 18 above)
Comment (cforkish) [reply apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:97]: our style guide isn't opinionated about this so in some style guide update proposals i'm working on i want to propose that we declare all class properties that don't need access to `this` in their initialization outside of the constructor. we only use constructors when necessary, so adopting this as a convention makes it easier to add or remove a constructor because you don't have to move any prop declarations around.

File: apps/src/templates/sectionProgress/sectionProgressLoader.js (hunk 17 above)
Comment (cforkish) [reply apps/src/templates/sectionProgress/sectionProgressLoader.js:112]: i don't want to trigger another build, but if the current build fails i will do this.

File: apps/src/templates/sectionProgress/sectionProgressConstants.js (hunk 16 above)
Comment (cforkish) [reply apps/src/templates/sectionProgress/sectionProgressConstants.js:22]: we have a lot of shape-auditing to do, so let's address this as part of that work (possibly when we add shape validation to redux?)

File: apps/src/templates/progress/progressHelpers.js (hunk 15 above)
Comment (cforkish) [reply apps/src/templates/progress/progressHelpers.js:223]: this method is actually deprecated and will be removed once the refactor is complete.

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx (hunk 18 above)
Comment (jamescodeorg) [reply apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:97]: Not sure how I feel about this as a general convention, but sounds good for this set of work.
